stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -616,166 +307,85 @@ private void releaseChildNodeRemovedEvent(ChildNodeRemovedEvent childNodeRemoved
 
-    public void removeWithEvent(int index) {
-      Object element = remove(index);
+    public void onChildNodeRemoved(
+        int parentNodeId,
+        int nodeId) {
       ChildNodeRemovedEvent removedEvent = acquireChildNodeRemovedEvent();
-      removedEvent.parentNodeId = mParentNodeId;
-      removedEvent.nodeId = mObjectIdMapper.getIdForObject(element);
+
+      removedEvent.parentNodeId = parentNodeId;
+      removedEvent.nodeId = nodeId;
       mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", removedEvent);
+
       releaseChildNodeRemovedEvent(removedEvent);
     }
-  }
 
-  private final class DOMObjectIdMapper extends ObjectIdMapper {
-    @Override
-    protected void onMapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+    public void onChildNodeInserted(
+        DocumentView view,
+        Object element,
+        int parentNodeId,
+        int previousNodeId,
+        Accumulator<Object> insertedElements) {
+      ChildNodeInsertedEvent insertedEvent = acquireChildNodeInsertedEvent();
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.hook(object);
-    }
+      insertedEvent.parentNodeId = parentNodeId;
+      insertedEvent.previousNodeId = previousNodeId;
+      insertedEvent.node = createNodeForElement(element, view);
 
-    @Override
-    protected void onUnmapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+      insertedElements.store(element);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeInserted", insertedEvent);
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.unhook(object);
+      releaseChildNodeInsertedEvent(insertedEvent);
     }
   }
 
-  private final class ProviderListener implements DOMProvider.Listener {
-    @Override
-    public void onPossiblyChanged() {
-      updateTree();
+  private final class PeerManagerListener extends DocumentAccessRegisteredListener {
+    public PeerManagerListener(Document document) {
+      super(document);
     }
 
     @Override
-    public void onAttributeModified(Object element, String name, String value) {
-      mDOMProvider.verifyThreadAccess();
-
-      AttributeModifiedEvent message = new AttributeModifiedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      message.value = value;
-      mPeerManager.sendNotificationToPeers("DOM.attributeModified", message);
-    }
-
-    @Override
-    public void onAttributeRemoved(Object element, String name) {
-      mDOMProvider.verifyThreadAccess();
-
-      AttributeRemovedEvent message = new AttributeRemovedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      mPeerManager.sendNotificationToPeers("DOM.attributeRemoved", message);
+    protected synchronized void onFirstPeerRegistered() {
+      super.onFirstPeerRegistered();
+      mDocument.addRef();
```

Comment:
my comment about `DocumentAccessRegisteredListener` ... this listener is deriving from that and then calls `mDocument.addRef` and `mDocument.release` anyway, so we end up calling these twice. I think we should just get rid of `DocumentAccessRegisteredListener`